feat(aci): issue alert action migration util for slack

iamrajjoshi opened the pull request:
this pr starts the implementation of migration util for issue alert migration to aci. i realized the pr would become a bit annoying to revidew, so wanted to get this out the main portion for some 👀 .

added some comments to the code for context.

ceorourke committed:
feat(aci): Add helpers to dual write to actions and dtaconditions

iamrajjoshi committed:
:wrench: chore: implementated migration util for slack

iamrajjoshi commented on src/sentry/workflow_engine/migration_helpers/rule_action.py:
```diff
@@ -0,0 +1,106 @@
+    """
+
+    # # If we have a specific blob type, we need to sanitize the action data to the blob type
+    if action_type == Action.Type.SLACK:
```
i tried out a mapping for this from action_type to the dataclass, but i liked having more explicit typing here since otherwise i think i would have to put a generic type.

iamrajjoshi replied:
open to better ideas

iamrajjoshi commented on src/sentry/workflow_engine/migration_helpers/rule_action.py:
```diff
@@ -0,0 +1,106 @@
+    Builds a SlackDataBlob from the action data.
+    Only includes the keys that are not None.
+    """
+    return SlackDataBlob(
```
narrowing the type that can exist in the data blob, should have something similar for most integrations, but might be a bit more difficult (& have diminishing gains) for sentry apps and such

iamrajjoshi replied:
update: i started defining separate data classes for each integration in stacked prs that depend on this. hopefully works out here.

i think we should move these builders somewhere else maybe so we can use it to validate for the actions endpoint.

iamrajjoshi commented on src/sentry/workflow_engine/migration_helpers/rule_action.py:
```diff
@@ -0,0 +1,106 @@
+        }
+
+
+def build_notification_actions_from_rule_data_actions(
```
main method that will be used

iamrajjoshi commented on src/sentry/workflow_engine/migration_helpers/rule_action.py:
```diff
@@ -0,0 +1,106 @@
+        notification_action = Action(
+            type=action_type,
+            data=(
+                # If the target type is specific, sanitize the action data
```
need to clean this portion up

iamrajjoshi commented on src/sentry/workflow_engine/migration_helpers/rule_action.py:
```diff
@@ -0,0 +1,106 @@
+            k: v
+            for k, v in action.items()
+            if k
+            not in [
```
convoluted, but basically popping keys we save elsewhere or don't want to save

iamrajjoshi committed:
:bug: fix: mypy and logging

iamrajjoshi commented on src/sentry/workflow_engine/typings/notification_action.py:
```diff
@@ -0,0 +1,113 @@
+- uuid: maps to action id
+- id: maps to action type
+"""
+EXCLUDED_ACTION_DATA_KEYS = ["uuid", "id"]
```
currently not saving the uuid since we are going to replace with action.id, but i am also considering saving it in the data blob incase something bad happens and we need it to recover? 🤔

cathteng commented on src/sentry/workflow_engine/migration_helpers/rule_action.py:
```diff
@@ -0,0 +1,135 @@
+        # For email, the target type is user
+        # For sentry app, the target type is sentry app
+        # FWIW, we don't use target type for issue alerts
+        target_type = ACTION_TYPE_2_TARGET_TYPE_RULE_REGISTRY.get(action_type)
```
we store this because the same workflow can be attached to metric alerts? what happens if the action_type doesn't exist in `ACTION_TYPE_2_TARGET_TYPE_RULE_REGISTRY`? maybe you could enforce that everything exists in a test?

iamrajjoshi replied:
we are planning to prevent the user from attaching issuer alert actions to metric alert actions (and vice versa) until we work on notification platform and making out notification handlers alert type agnostic.

until then, we can assume these actions we migrate will only be used for the errors detector we are creating.

technically, because of that, we don't even need to maintain this column and since its nullable, i remove it.

cathteng commented on src/sentry/workflow_engine/migration_helpers/rule_action.py:
```diff
@@ -0,0 +1,135 @@
+            # Get the target_identifier if it exists
+            target_identifier_key = ACTION_TYPE_2_TARGET_IDENTIFIER_KEY.get(action_type)
+            if target_identifier_key is not None:
+                target_identifier = action.get(target_identifier_key)
```
what happens if this doesn't exist?

cathteng commented on src/sentry/workflow_engine/migration_helpers/rule_action.py:
```diff
@@ -0,0 +1,135 @@
+                # Otherwise, use the action data as is
+                sanitize_to_action(action, action_type)
+                if target_type == ActionTarget.SPECIFIC
+                else action
```
we put the entire old blob in here?

cathteng commented on src/sentry/workflow_engine/migration_helpers/rule_action.py:
```diff
@@ -0,0 +1,135 @@
+            target_type=target_type,
+        )
+
+        notification_action.save()
+
+        notification_actions.append(notification_action)
```
instead of saving immediately, you could create a list of unsaved objects and then call `bulk_create` https://docs.djangoproject.com/en/5.1/ref/models/querysets/#bulk-create, but this won't call the `save()` method

cathteng reviewed (commented):
i assume we are populating all the fields for the action in case a workflow is used for the error detector and/or metric detectors?
my main concern is that we should be really careful about assuming things exist in dictionaries and have a fallback plan if things don't exist

iamrajjoshi committed:
:recycle: ref: remove target_type attribute usage

iamrajjoshi commented on src/sentry/workflow_engine/typings/notification_action.py:
```diff
@@ -0,0 +1,129 @@
+
+
+@dataclass
+class DataBlob:
```
kept this here incase we want a mapping instead of the if statements i had above. i liked having intellisense on what type i am building so i don't have a mapping, but we can put this in as well.

malwilley committed:
chore(search-query-builder): Remove flag definitions (#83038)

saponifi3d commented on src/sentry/workflow_engine/migration_helpers/alert_rule.py:
```diff
@@ -1,20 +1,152 @@
+        )
+        if not integration:
+            return None
+        for type in Action.Type.choices:
+            if type[0] == integration.provider:
+                return type[1]
```
nit: since this is all static, could we convert this to a dict at module import, then just do a dict lookup by provider?

```python
PROVIDER_LOOKUP = dict(Action.Type.choices)
```

Then this could just be:
```suggestion
        try:
            return PROVIDER_LOOKUP[integration.provider]
        except KeyError:
            return None
```

saponifi3d commented on src/sentry/workflow_engine/migration_helpers/alert_rule.py:
```diff
@@ -1,20 +1,152 @@
+    threshold_to_condition = {
+        AlertRuleThresholdType.ABOVE.value: Condition.GREATER,
+        AlertRuleThresholdType.BELOW.value: Condition.LESS,
+        # TODO add ABOVE_AND_BELOW for anomaly detection
```
this one could be tricky, we may want to just make an `AnomalyDetectionCondition` which makes the request to seer - since the comparison value is dynamic as well.

saponifi3d commented on src/sentry/workflow_engine/migration_helpers/alert_rule.py:
```diff
@@ -1,20 +1,152 @@
+    }
+
+    data_condition_group = alert_rule_detector.detector.workflow_condition_group
+    if not data_condition_group:
+        logger.warning(
+            "Could not find data_condition_group",
+            extra={"detector_id": alert_rule_detector.id},
+        )
+        return None
```
should this just create the data condition group if it doesn't exist? (since this is a migration helper)

saponifi3d reviewed (commented):
i think it'd help to break down this PR a bit.

A good breakdown might be:
- Migration helpers / alert_rule
   - with applicable types
- Migration helpers / rule_action
   - with applicable types

It's not just for the ease of review, but also the ease of merging / maintenance as well. It keeps the chunks of work together so we can easily revert if needed (for whatever reason). this also helps with keeping PRs up to date and merge conflicts to a minimum.

ArthurKnaus committed:
fix(org-stats): Require project membership (#83085)

Show fallback UI if no projects are available.

Close https://github.com/getsentry/sentry/issues/82998

priscilawebdev committed:
ref(quick-start): Remove not used feature flag (#83001)

ArthurKnaus committed:
feat(onboarding): Generic copy DSN (#83088)

Move the "copy DSN" section into manual configuration.
Add it to every platform doc that instructs to use the wizard.

Closes https://github.com/getsentry/sentry/issues/80179

asottile-sentry committed:
ref: remove attachment_id column part 2 (#83036)

<!-- Describe your PR here. -->

asottile-sentry committed:
ref: upgrade grpcio (#83041)

for 3.13

<!-- Describe your PR here. -->

asottile-sentry committed:
ref: upgrade pydantic to 1.10.20 (#83040)

this adds python 3.13 support

<!-- Describe your PR here. -->

BYK committed:
feat: Enable Django cache_spans in Sentry SDK (#82645)

This is a follow up to
[getsentry/sentry-python#3300](https://github.com/getsentry/sentry-python/issues/3300)
where we think we have addressed the performance issues regarding
`cache_spans` in `DjangoIntegration`. This is a test and if we can
confirm the lack of perf regression, we will enable this option by
default in Sentry SDK as the next step.

nikkikapadia committed:
chore(dashboards): Remove edit access FF code from backend (#83042)

Removing the edit access feature flag from the backend. This makes all
of the edit access permissions code the default now. Also adjusted the
tests for this.

Context: https://sentry.slack.com/archives/C01MYDW8Y7K/p1736203608171589

Next step: Removing the feature flag from the options repo.

narsaynorath committed:
fix(dashboard): Fix exports in Add to Dashboard modal (#83046)

Exporting these files from the `widgetCard/index.tsx` file was causing
the following error: `Cannot access 'WidgetCardPanel' before
initialization`

This was from a circular import and the fix is to move these
dependencies into their own files which don't rely on other imports that
may trigger the issue.

shruthilayaj committed:
chore(alerts): Use RPC on alert details page (#83051)

Use RPC if EAP query on alert details

markstory committed:
chore: Cleanup old TODOs (#82201)

Some of these TODOs have been in the code base for several years and are
not something I'll be able to address anytime soon, nor do I still
believe they need to be addressed. The backup related TODOs are past
their expiry date and can be cleaned up.

gggritso committed:
fix(dashboards): Allow sorting Dashboards table by release (#82943)

Closes https://github.com/getsentry/sentry/issues/82925. Somewhere along
the line, the logic that sorts data by release on the frontend (due to
backend dataset limitation) broke. If someone created a widget, sorted
by release, and _saved it_, it works. But, when someone is _switching
to_ sorting by release the preview is busted because we forgot to fetch
releases for them.

This PR makes sure that if they update widget settings to start sorting
by release, we fetch the releases.

gggritso committed:
docs(ui): Add more `GlobalDrawer` examples (#83093)

Expanding the documentation with more examples and details of how it
works.

tustanivsky committed:
Update UE SDK onboarding wizard (#82995)

This PR enhances the UE SDK onboarding wizard with the following
changes:

1. Updated Installation Instructions:
- Clarified which artifact users should download from the [plugin's
GitHub Releases
page](https://github.com/getsentry/sentry-unreal/releases).
- Added a link to [alternate installation
methods](https://docs.sentry.io/platforms/unreal/#install).
2. Improved Crash Reporter guidance by adding an explicit notice that it
should be used for automatic crash capturing with UE versions older than
5.2.
3. Removed the redundant info about including debug files in packaged
game build.

For more details, refer to [issue
#82610](https://github.com/getsentry/sentry/issues/82610).

untitaker committed:
chore(snuba): Fix tests for rust errors processor (#83031)

blocked on https://github.com/getsentry/snuba/pull/6720

nikkikapadia committed:
chore(dashboards): Remove Edit Access feature flag references (#83021)

Removes all instances of the edit access feature flag in the code base.
Also had to make some test adjustments to handle this.

Context: https://sentry.slack.com/archives/C01MYDW8Y7K/p1736203608171589

Next step: Removing the feature flag from the options repo.

wmak committed:
fix(discover): Handles when issue is in top events with transactions (#83032)

- Sometimes queries are being run on the transactions dataset with the
issues field, but since issue doesn't exist there its causing the query
to fail
- Fixes SENTRY-3DPS

roggenkemper committed:
chore(issue-details): Only make one request to readable tags (#83061)

this pr updates the calls to `useGroupTagsReadable` so that we only make
one call, instead of 2

roggenkemper committed:
fix(issue-details): Use correct environments value for tag drawer (#83063)

this pr fixes an issue with the environment value in the tag drawer. we
already made a call for tags in the search bar, but because the tag
drawer had a slightly different way of deriving the value of
`environment`, we couldn't take advantage of the caching. now, the tag
drawer should instantly load on the issue details page.

roggenkemper committed:
chore(issue-details): Fix mismatch in tag distributions with same value (#83074)

this pr fixes a minor issue where two tags could show the same value but
their tag distributions looked different because we weren't rounding the
numbers


before: 
![Screenshot 2025-01-07 at 4 09
46 PM](https://github.com/user-attachments/assets/e6c6cc57-80c3-47c1-bb19-eec184841cea)

mrduncan committed:
chore(issues): Clean up escalating-issues-v2 (#82495)

This logging was only enabled for the Sentry organization, we're no
longer using it, and it seems to be broken:
[SENTRY-2BCX](https://sentry.sentry.io/issues/4742297960/).
> InvalidParams
> Raw metric errors.event_ingested does not exist

The issue above shows this code being executed:

https://github.com/getsentry/sentry/blob/a2dc09f6529fffed9e6227a76ded966c57d857e7/src/sentry/issues/escalating.py#L222-L244


https://github.com/search?q=org%3Agetsentry+escalating-issues-v2+NOT+repo%3Agetsentry%2Fsentry-mirror-suggested-fix+NOT+repo%3Agetsentry%2Fsuggested-fix-pr-demo&type=code

https://github.com/getsentry/getsentry/pull/16004 disables all Sentry
orgs from using this feature.

malwilley committed:
fix(issue-stream): Wait until stats request has finished before saving groups into cache (#83073)

Fixes a bug in the refactored component where navigating to an issue and
going back to the stream would never load the stats. This is because I
started caching when the request completed instead of when the component
unmounted, but never waited for the stats request to complete.

- Makes request async and use `requestPromise` instead of `request`
- Wait for request to finish before caching
- Use `api.clear()` instead of `lastRequestRef.current.cancel()`

ceorourke committed:
chore(alerts): Remove activated alerts subscription to projects (#83060)

Remove the code for activated alerts to subscribe to projects and put
back what we were using before. I can't remove
`conditionally_subscribe_project_to_alert_rules` and
`subscribe_projects` just yet because some tests rely on them, but in
another one I will remove those as well and all the references.

roggenkemper committed:
feat(issue-details): Hide tag preview on search (#83055)

this pr hides the tag preview on issue details if the user performs a
search. right now the tag preview doesn't update based on the search.
even if it did, it would be a little weird to see the tags we choose to
show be updated based off the search, so if a search does happen, we
just hide the preview.

snigdhas committed:
feat(metric-issues): Add detector details to sidebar (#82972)

![image](https://github.com/user-attachments/assets/a0b0b13d-75ae-4029-9daa-7ff236e20580)

roggenkemper committed:
fix(issue-details): Fix tags preview if no tags (#83069)

this pr fixes a bug where we showed a preview of nothing instead of
showing nothing if there are no tags.

roggenkemper committed:
fix(issue-details): Disable tags button if no tags (#83070)

this pr disabled the "All Tags" button if there are no tags

MichaelSun48 committed:
fix(issue-views): Fix bug where unable to rename fresh duplicated view (#83094)

Fixes an issue where after users were unable to rename a tab right after
it's been duplicated (see this
[replay](https://sentry.sentry.io/replays/4290855474744580b0bd4bd264794273/?t=248.437&t_main=breadcrumbs),
3:30 for behavior)

This was happening because the tabListState's selected key wasn't being
set to the duplicated tab after it had been created, causing the active
element to still be on the old tab:


![image](https://github.com/user-attachments/assets/3e665f76-ec0e-4e66-90c0-e5023aaa912c)
^ `document.activeElement` is being highlighted here, it is erroneously
on the original tab, and not on the duplicated tab, despite the
duplicated tab appearing to be selected.


fixes JAVASCRIPT-2XA4

iamrajjoshi committed:
:recycle: ref: surface slack rate limit errors (#83068)

iamrajjoshi committed:
:bug: fix(msteams): enable personal notification settings (#82983)

aliu39 committed:
chore(replay): add a generic log sampling filter and sample replay ingest logs (#83049)

scttcper committed:
feat(issues): Split out issue event navigation, fix preloading (#83053)

michellewzhang committed:
ref(flags): limit scopes for secret updates (#82897)

closes https://github.com/getsentry/team-replay/issues/522

changes the secret endpoint permissions so that only managers & owners
can update a secret (anyone can post) -- only exception is the original
creator of the secret can always update their secret, regardless of
their scope.

lobsterkatie committed:
fix(code mappings): Handle non-int project ids (#83087)

It's unclear how this happens (user error? misconfigured integration?), but from time to time our code mappings endpoint gets hit with a request where the project id is a string or a hex value rather than an integer. This adds handling for that case, returning a 400 response rather than simply erroring out.

JonasBa committed:
ref(trace) fix overlapping labels (#83035)

Partially mitigate overlapping web vitals indicators.

---------

Co-authored-by: Abdullah Khan <abdullahkhan@PG9Y57YDXQ.local>

GabeVillalobos committed:
fix(hybrid-cloud): Adds option to skip invalid audit log events by event id (#82820)

MichaelSun48 committed:
fix(issue-views): Fix issue where new views do not have a query saved (#83099)

Fixes an issue where clicking on a query suggestion in the new view page
would not populate the search bar with the query.

JoshFerge committed:
feat(uptime): produce to new snuba topic in check results consumer (#82513)

part of https://github.com/getsentry/team-uptime/issues/23

requires https://github.com/getsentry/sentry-kafka-schemas/pull/359 and
https://github.com/getsentry/snuba/pull/6697

evanpurkhiser committed:
feat(ui): Extract generalized "CheckInTimeline" component (#83101)

This will be used for both crons and uptime

Will add a story for this component and some of the various additional
components soon after

iamrajjoshi committed:
:wrench: chore(workflow engine): remove legacy_notification_type column (#83100)

lobsterkatie committed:
chore(seer grouping): Add debug logging for old excess frames check (#83082)

As of https://github.com/getsentry/sentry/pull/82414, we should no longer be using `TooManyOnlySystemFramesException`, but the fact that we're still getting `did_call_seer` metrics tagged with `"over-threshold-frames"` says that we still are, somehow. This is currently a blocker on merging https://github.com/getsentry/sentry/pull/82434, which removes `TooManyOnlySystemFramesException` and the code associated with it, so it'd be nice to figure out why it's happening so we can fix it.

Try as I might, I can't replicate it, nor can I find the flaw in the logic, so this adds a temporary log to try to help figure out how in the heck it's happening. It also adds a `record_metrics` parameter to the `has_too_many_contributing_frames` helper, so that I can call it for purposes of the logging without messing up the metrics it collects. Once I figure it out what's happening, I'll back out all of these changes.

ameliahsu committed:
feat(workflow-engine): add `IssueCategoryConditionHandler` (#83056)

add `IssueCategoryConditionHandler`

evanpurkhiser committed:
ref(insights): Add basic cron overview to insights (#83029)

evanpurkhiser committed:
ref(alerts): Consistent bread crumb labels (#83020)

Instead of using the name of the thing, which is displayed directly in
the header below, use the name of the alert type we're looking at.

mifu67 commented:
lgtm
